### PR TITLE
Pico proxy as nodelet

### PR DESCRIPTION
**Pull request by marinagmoreira**

### CODE REVIEW

**Comment by marinagmoreira** [comment]
Tested local, awaiting robot test

**Comment by marinagmoreira** [comment]
Tested in the lab

**Comment by marinagmoreira** [comment]
+ platform submodule update to working rootfs

**Comment by marinagmoreira** [comment]
Tested on the robot

File: hardware/pico_driver/src/pico_proxy.cc [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+// Messages
+#include <sensor_msgs/Image.h>
+#include <ff_msgs/PicoflexxIntermediateData.h>
+
+#include <string>
+#include <map>
+
+
+// Layer extraction
+std::vector <cv::Mat> layers(4);
```
Comment (kbrowne15) [inline hardware/pico_driver/src/pico_proxy.cc:44]: Is there a reason this isn't a private data member?

File: hardware/pico_driver/src/pico_proxy.cc [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+class PicoProxyNodelet : public ff_util::FreeFlyerNodelet  {
+ public:
+  PicoProxyNodelet() : ff_util::FreeFlyerNodelet(false) {}
+  ~PicoProxyNodelet() {}
+
+ protected:
+  void Initialize(ros::NodeHandle *nh) {
+    std::string topic, topic_type;
+    if (!ros::param::get("topic", topic))
+      ROS_FATAL("You need to pass a topic to the pico proxy");
```
Comment (kbrowne15) [inline hardware/pico_driver/src/pico_proxy.cc:60]: Not a big deal but you can change these to NODELET_FATAL, etc.

File: hardware/pico_driver/src/pico_proxy.cc [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+class PicoProxyNodelet : public ff_util::FreeFlyerNodelet  {
+ public:
+  PicoProxyNodelet() : ff_util::FreeFlyerNodelet(false) {}
+  ~PicoProxyNodelet() {}
+
+ protected:
+  void Initialize(ros::NodeHandle *nh) {
+    std::string topic, topic_type;
+    if (!ros::param::get("topic", topic))
```
Comment (kbrowne15) [inline hardware/pico_driver/src/pico_proxy.cc:59]: Again, can we avoid ros params :)

**Review by kbrowne15 — APPROVED** [review]
Looks really good. I just have a few small changes.

File: astrobee/launch/robot/MLP.launch [hunk 4]
```diff
@@ -400,7 +400,19 @@
      <arg name="default" value="true" />
    </include>
 
-  <include file="$(find pico_driver)/launch/pico_proxy.launch" />
+   <param name="topic" type="str" value="/hw/depth_haz/extended" />
```
[the remark this answers is not in the document]
Comment (marinagmoreira) [reply astrobee/launch/robot/MLP.launch:403]: I think the pico_proxy tool is made to work with any depth camera and for type extended and depth_image. If we set it with ff_names then we're making it static

File: hardware/pico_driver/src/pico_proxy.cc (hunk 1 above)
Comment (marinagmoreira) [reply hardware/pico_driver/src/pico_proxy.cc:44]: It crashed if I made it private :/ I don't know why

File: astrobee/launch/robot/MLP.launch (hunk 4 above)
[the remark this answers is not in the document]
Comment (kbrowne15) [reply astrobee/launch/robot/MLP.launch:403]: Could we put it in a lua config instead?

File: hardware/pico_driver/src/pico_proxy.cc (hunk 1 above)
Comment (kbrowne15) [reply hardware/pico_driver/src/pico_proxy.cc:44]: Weird, okay let's leave it alone then.

File: astrobee/launch/robot/MLP.launch (hunk 4 above)
[the remark this answers is not in the document]
Comment (marinagmoreira) [reply astrobee/launch/robot/MLP.launch:403]: How would we deal with multiple instances on lua?

[the remark this answers is not in the document]
Comment (kbrowne15) [reply astrobee/launch/robot/MLP.launch:403]: Oh, I see what you are saying. Is there any way we can do tricks like what we do for the nav and dock cam or somehow use the name of the node to read in a different lua config file?

[the remark this answers is not in the document]
Comment (marinagmoreira) [reply astrobee/launch/robot/MLP.launch:403]: I guess so, I'll check what we do with the cameras :)

**Review by bcoltin — CHANGES_REQUESTED** [review]
I agree with Katy that we should try to get rid of the global variable. If it is crashing we should try to understand why. Another option is to make it a static local variable so it won't be reinitialized.

This can be a separate pull request (since this one doesn't make it worse), but if this node is taking any significant CPU, I expect we could reduce it some with the copies in the extendedcallback.

**Comment by marinagmoreira** [comment]
> I agree with Katy that we should try to get rid of the global variable. If it is crashing we should try to understand why. Another option is to make it a static local variable so it won't be reinitialized.

Actually I was just initializing it wrong, fixed it xD

**Review by bcoltin — APPROVED** [review]

**Pull request merged**